File: src/main/java/com/github/kindrat/cassandra/client/service/TableContext.java
package com.github.kindrat.cassandra.client.service;

import com.datastax.driver.core.*;
import com.github.kindrat.cassandra.client.ui.DataObject;
import com.github.kindrat.cassandra.client.util.EvenMoreFutures;
import com.google.common.util.concurrent.ListenableFuture;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.control.TableColumn;
import lombok.Getter;
import lombok.Synchronized;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static com.github.kindrat.cassandra.client.ui.fx.TableColumns.buildColumn;
import static com.github.kindrat.cassandra.client.util.EvenMoreFutures.*;
import static com.github.nginate.commons.lang.NStrings.format;

@Slf4j
public class TableContext {

    private final Map<Integer, String> pagingStates = new HashMap<>();
    @Getter
    private final String table;
    private final String query;
    @Getter
    private final TableMetadata tableMetadata;
    @Getter
    private final List<TableColumn<DataObject, Object>> columns;
    private final CassandraClientAdapter clientAdapter;
    private final int pageSize;

    private int page;
    private CompletableFuture<ObservableList<DataObject>> currentPage;
    private volatile ResultSet resultSet;

    private TableContext(String table, String query, TableMetadata tableMetadata, CassandraClientAdapter clientAdapter,
            int pageSize) {
        this.table = table;
        this.tableMetadata = tableMetadata;
        this.clientAdapter = clientAdapter;
        this.pageSize = pageSize;
        this.query = query;
        columns = buildColumns();
    }

    public static TableContext fullTable(String table, TableMetadata tableMetadata,
            CassandraClientAdapter clientAdapter, int pageSize) {
        return new TableContext(table, "select * from " + table, tableMetadata, clientAdapter, pageSize);
    }

    public static TableContext customView(String table, String query, TableMetadata tableMetadata,
            CassandraClientAdapter clientAdapter, int pageSize) {
        return new TableContext(table, query, tableMetadata, clientAdapter, pageSize);
    }

    @Synchronized
    public CompletableFuture<ObservableList<DataObject>> nextPage() {
        if (!hasNextPage()) {
            return currentPage;
        }
        page++;
        currentPage = toCompletable(resultSet.fetchMoreResults())
                .thenApply(this::parseResultSet)
                .thenApply(FXCollections::observableList);
        return currentPage;
    }

    @Synchronized
    public boolean hasPreviousPage() {
        return page > 0;
    }

    @Synchronized
    public boolean hasNextPage() {
        return !resultSet.isExhausted();
    }

    @Synchronized
    public CompletableFuture<ObservableList<DataObject>> previousPage() {
        page = Math.max(0, page - 1);
        currentPage = loadPage(page);
        return currentPage;
    }

    @Synchronized
    private CompletableFuture<ObservableList<DataObject>> loadPage(int page) {
        log.debug("Loading page {}", page);
        Statement statement = getStatement();
        String rawPagingState = pagingStates.get(page);
        if (rawPagingState != null) {
            PagingState pagingState = PagingState.fromString(rawPagingState);
            log.debug("Loading page {} with pagination state {}", page, rawPagingState);
            statement.setPagingState(pagingState);
        }
        return clientAdapter.executeStatement(statement)
                .thenApply(result -> {
                    ExecutionInfo executionInfo = result.getExecutionInfo();
                    executionInfo.getWarnings().forEach(log::warn);
                    printQueryTrace(page, executionInfo.getQueryTraceAsync());
                    this.resultSet = result;
                    return result;
                })
                .thenApply(this::parseResultSet)
                .thenApply(FXCollections::observableList)
                .whenComplete(logErrorIfPresent(format("Page {} load failed", page)));
    }

    private Statement getStatement() {
        Statement statement = new SimpleStatement(query);
        statement.setFetchSize(pageSize);
        return statement.enableTracing();
    }

    private List<DataObject> parseResultSet(ResultSet resultSet) {
        AtomicInteger start = new AtomicInteger(pageSize * page);
        int availableWithoutFetching = resultSet.getAvailableWithoutFetching();
        int currentPageSize = Math.min(availableWithoutFetching, pageSize);
        PagingState nextPageState = resultSet.getExecutionInfo().getPagingState();
        log.debug("Page {} size {}", page, currentPageSize);
        if (nextPageState != null) {
            String value = nextPageState.toString();
            log.debug("Page {} paging state {}", page + 1, value);
            pagingStates.put(page + 1, value);
        } else {
            log.info("No pagination state for remaining entries. Fetching them within current page.");
            currentPageSize = availableWithoutFetching;
        }
        return StreamSupport.stream(resultSet.spliterator(), false)
                .limit(currentPageSize)
                .map(row -> parseRow(start.getAndIncrement(), row))
                .collect(Collectors.toList());
    }

    private DataObject parseRow(int index, Row row) {
        DataObject dataObject = new DataObject(index);
        for (ColumnDefinitions.Definition definition : row.getColumnDefinitions()) {
            TypeCodec<Object> codec = CodecRegistry.DEFAULT_INSTANCE.codecFor(definition.getType());
            dataObject.set(definition.getName(), row.get(definition.getName(), codec));
        }
        return dataObject;
    }

    private List<TableColumn<DataObject, Object>> buildColumns() {
        List<TableColumn<DataObject, Object>> columns = new ArrayList<>();

        TableColumn<DataObject, Object> counterColumn = buildColumn(DataType.cint(), "#");
        counterColumn.setCellValueFactory(param -> {
            Integer object = param.getValue().getPosition();
            return new SimpleObjectProperty<>(object);
        });
        counterColumn.setEditable(false);

        columns.add(counterColumn);

        tableMetadata.getColumns().forEach(columnMetadata -> {
            DataType type = columnMetadata.getType();
            TableColumn<DataObject, Object> column = buildColumn(type, columnMetadata.getName());
            column.setCellValueFactory(param -> {
                Object object = param.getValue().get(columnMetadata.getName());
                return new SimpleObjectProperty<>(object);
            });
            column.setOnEditCommit(event -> {
                log.debug("Updating row value with {}", event.getRowValue());
                clientAdapter.update(tableMetadata, event).whenComplete(loggingConsumer(aVoid -> event.getRowValue()));
            });
            columns.add(column);
        });
        return columns;
    }

    private void printQueryTrace(int page, ListenableFuture<QueryTrace> listenableFuture) {
        EvenMoreFutures.toCompletable(listenableFuture)
                .whenCompleteAsync((trace, throwable) -> {
                    if (trace != null) {
                        log.info("Executed page {} query trace : [{}] " +
                                        "\n\t started at {} and took {} μs " +
                                        "\n\t coordinator {}" +
                                        "\n\t request type {}" +
                                        "\n\t parameters {}" +
                                        "\n\t events",
                                page, trace.getTraceId(), trace.getStartedAt(), trace.getDurationMicros(),
                                trace.getCoordinator(), trace.getRequestType(), trace.getParameters(),
                                trace.getEvents());
                    }
                    if (throwable != null) {
                        log.error("Query trace could not be read", throwable);
                    }
                });

    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/service/TableContext.kt
package com.github.kindrat.cassandra.client.service

import com.datastax.oss.driver.api.core.cql.AsyncResultSet
import com.datastax.oss.driver.api.core.cql.QueryTrace
import com.datastax.oss.driver.api.core.cql.Row
import com.datastax.oss.driver.api.core.cql.SimpleStatement
import com.datastax.oss.driver.api.core.metadata.schema.ColumnMetadata
import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata
import com.datastax.oss.driver.api.core.type.DataTypes
import com.datastax.oss.driver.api.core.type.codec.registry.CodecRegistry
import com.github.kindrat.cassandra.client.ui.DataObject
import com.github.kindrat.cassandra.client.ui.fx.TableColumns
import com.github.kindrat.cassandra.client.util.EvenMoreFutures.logErrorIfPresent
import com.github.kindrat.cassandra.client.util.EvenMoreFutures.loggingConsumer
import javafx.beans.property.SimpleObjectProperty
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.scene.control.TableColumn
import mu.KotlinLogging
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Function
import java.util.stream.StreamSupport
import kotlin.math.max
import kotlin.math.min
import kotlin.streams.toList

private val logger = KotlinLogging.logger {}

class TableContext private constructor(
        val table: String,
        private val query: String,
        val tableMetadata: TableMetadata,
        private val clientAdapter: CassandraClientAdapter,
        private val pageSize: Int) {

    private val pagingStates: MutableMap<Int, ByteArray> = HashMap()
    val columns: List<TableColumn<DataObject, Any>> = buildColumns()
    private var page = 0
    private var currentPage: CompletionStage<ObservableList<DataObject>>? = null
    private val resultSet: AtomicReference<AsyncResultSet?> = AtomicReference(null)

    @Synchronized
    fun nextPage(): CompletionStage<ObservableList<DataObject>> {
        if (!hasNextPage()) {
            return currentPage ?: CompletableFuture.completedStage(FXCollections.emptyObservableList())
        }
        page++
        currentPage = resultSet.get()?.fetchNextPage()
                ?.thenApply { parseResultSet(it) }
                ?.thenApply { FXCollections.observableList(it) }
        return currentPage ?: CompletableFuture.completedStage(FXCollections.emptyObservableList())
    }

    @Synchronized
    fun hasPreviousPage(): Boolean {
        return page > 0
    }

    @Synchronized
    fun hasNextPage(): Boolean {
        return resultSet.get()?.hasMorePages() ?: false
    }

    @Synchronized
    fun previousPage(): CompletionStage<ObservableList<DataObject>> {
        page = max(0, page - 1)
        val loadedPage = loadPage(page)
        currentPage = loadedPage
        return loadedPage
    }

    @Synchronized
    private fun loadPage(page: Int): CompletionStage<ObservableList<DataObject>> {
        logger.debug("Loading page $page")
        val statement = statement
        val rawPagingState = pagingStates[page]
        if (rawPagingState != null) {
            val pagingState = ByteBuffer.wrap(rawPagingState)
            logger.debug("Loading page $page with pagination state $rawPagingState")
            statement.pagingState = pagingState
        }
        return clientAdapter.executeStatement(statement)
                .thenApply {
                    val executionInfo = it.executionInfo
                    executionInfo.warnings.forEach { message -> logger.warn(message) }
                    printQueryTrace(page, executionInfo.queryTraceAsync)
                    resultSet.set(it)
                    it
                }
                .thenApply { parseResultSet(it) }
                .thenApply { FXCollections.observableList(it) }
                .whenComplete(logErrorIfPresent<ObservableList<DataObject>, Throwable>("Page $page load failed"))
    }

    private val statement: SimpleStatement
        get() = SimpleStatement.newInstance(query)
                .setTracing(true)
                .setPageSize(pageSize)

    private fun parseResultSet(resultSet: AsyncResultSet): List<DataObject> {
        val start = AtomicInteger(pageSize * page)
        val availableWithoutFetching = resultSet.remaining()
        var currentPageSize = min(availableWithoutFetching, pageSize)
        val nextPageState = resultSet.executionInfo.pagingState
        logger.debug("Page $page size $currentPageSize")
        if (nextPageState != null) {
            val value = nextPageState.array()
            logger.debug("Page ${page + 1} paging state $value")
            pagingStates[page + 1] = value
        } else {
            logger.info("No pagination state for remaining entries. Fetching them within current page.")
            currentPageSize = availableWithoutFetching
        }
        return StreamSupport.stream(resultSet.currentPage().spliterator(), false)
                .limit(currentPageSize.toLong())
                .map { parseRow(start.getAndIncrement(), it) }
                .toList()
    }

    private fun parseRow(index: Int, row: Row): DataObject {
        val dataObject = DataObject(index)
        for (definition in row.columnDefinitions) {
            val codec = CodecRegistry.DEFAULT.codecFor<Any>(definition.type)
            dataObject[definition.name.toString()] = row.get(definition.name, codec)!!
        }
        return dataObject
    }

    private fun buildColumns(): List<TableColumn<DataObject, Any>> {
        val columns: MutableList<TableColumn<DataObject, Any>> = ArrayList()
        val counterColumn = TableColumns.buildColumn<DataObject>(DataTypes.INT, "#")
        counterColumn.setCellValueFactory { SimpleObjectProperty(it.value.position) }
        counterColumn.isEditable = false
        columns.add(counterColumn)
        tableMetadata.columns.values.forEach { metadata: ColumnMetadata ->
            val type = metadata.type
            val column = TableColumns.buildColumn<DataObject>(type, metadata.name.toString())
            column.setCellValueFactory { SimpleObjectProperty(it.value[metadata.name.toString()]) }
            column.setOnEditCommit {
                logger.debug("Updating row value with ${it.rowValue}")
                clientAdapter.update(tableMetadata, it)
                        .whenComplete(loggingConsumer(Function<Void, Any> { _ -> it.rowValue }))
            }
            columns.add(column)
        }
        return columns
    }

    private fun printQueryTrace(page: Int, resultFuture: CompletionStage<QueryTrace>) {
        resultFuture.whenCompleteAsync { trace: QueryTrace?, throwable: Throwable? ->
            trace?.let {
                logger.info("""Executed page $page query trace : [${it.tracingId}] 
                                 started at ${it.startedAt} and took ${it.durationMicros} μs 
                                 coordinator ${it.coordinatorAddress}
                                 request type ${it.requestType}
                                 parameters ${it.parameters}
                                 events ${it.events}""")
            }
            throwable?.let { logger.error("Query trace could not be read", it) }
        }
    }

    companion object {
        @JvmStatic
        fun fullTable(table: String, tableMetadata: TableMetadata, clientAdapter: CassandraClientAdapter,
                      pageSize: Int): TableContext {
            return TableContext(table, "select * from $table", tableMetadata, clientAdapter, pageSize)
        }

        @JvmStatic
        fun customView(table: String, query: String, tableMetadata: TableMetadata,
                       clientAdapter: CassandraClientAdapter, pageSize: Int): TableContext {
            return TableContext(table, query, tableMetadata, clientAdapter, pageSize)
        }
    }
}